### PR TITLE
feat: introduce LOD rendering of torus geometries to improve quality of certain primitives

### PR DESCRIPTION
**Pull request by larsmoa**
Before:
image
After:
image
(notice the silhouette of the big pipe)

### CODE REVIEW

File: viewer/src/utilities/three/BoundingBoxLOD.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
+/*!
+ * Copyright 2021 Cognite AS
+ */
+import * as THREE from 'three';
+
+export class BoundingBoxLOD extends THREE.Object3D {
```
Comment (christjt) [inline viewer/src/utilities/three/BoundingBoxLOD.ts:6]: Really smart!

File: viewer/src/utilities/three/BoundingBoxLOD.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+          levels[i].object.visible = true;
+          this._currentLevel = i;
+        } else {
+          break;
+        }
+      }
+
+      for (let i = this._currentLevel + 1; i < levels.length; ++i) {
+        levels[i].object.visible = false;
+      }
```
Comment (christjt) [inline viewer/src/utilities/three/BoundingBoxLOD.ts:65]: If you reverse the sort order I think you can do this. This also assumes that all levels are initialized as `visible=false`

```js
    levels[this._currentLevel].visible = false;
    this._currentLevel = levels.findIndex(p => distanceToCamera >= p.distance);
    levels[this._currentLevel].visible = false;
```

Which looks cleaner in my opinion.

**Review by christjt — CHANGES_REQUESTED** [review]
Overall looks really good. Clever to extend an object3D to create a LOD node.

File: viewer/src/utilities/three/BoundingBoxLOD.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+  private readonly _updateVars = {
+    camPos: new THREE.Vector3(),
+    bounds: new THREE.Box3()
+  };
+  update(camera: THREE.Camera) {
+    const levels = this._levels;
+    const { camPos, bounds } = this._updateVars;
+    bounds.copy(this._boundingBox).applyMatrix4(this.matrixWorld);
+    const cameraZoom = camera instanceof THREE.PerspectiveCamera ? camera.zoom : 1.0;
```
[the remark this answers is not in the document]
Comment (larsmoa) [reply viewer/src/utilities/three/BoundingBoxLOD.ts:44]: We could, but this will just work - no reason to not have it work on non-perspective?

**Review by christjt — APPROVED** [review]

**Pull request merged**